s4e-backend/src/main/java/pl/cyfronet/s4e/controller/PasswordController.java

```diff
@@ -0,0 +1,91 @@
+
+@RestController
+@RequestMapping(API_PREFIX_V1)
+@RequiredArgsConstructor
+public class PasswordController {
+    private final AppUserService appUserService;
+    private final PasswordService passwordService;
+    private final ApplicationEventPublisher eventPublisher;
+
+    @ApiOperation("Create and send a password reset token based on email")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "Email with a password reset token was sent provided user with it existed")
+    })
+    @PostMapping(value = "/token-create")
+    public ResponseEntity<?> resetPassword(@RequestParam @Email @NotEmpty @Valid  String email) {
+        val optionalAppUser = appUserService.findByEmail(email);
+        if (optionalAppUser.isPresent()) {
+            eventPublisher.publishEvent(new OnPasswordResetTokenEmailEvent(optionalAppUser.get(), LocaleContextHolder.getLocale()));
+        }
+        return ResponseEntity.ok().build();
+    }
+
+    @ApiOperation("Validate if reset token was used or expired")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "Password reset token is valid"),
+            @ApiResponse(code = 401, message = "The token has expired"),
+            @ApiResponse(code = 404, message = "The token was not found")
+    })
+    @GetMapping(value = "/token-validate")
+    public ResponseEntity<?> validateToken(@RequestParam @NotEmpty @Valid String token) throws NotFoundException, PasswordResetTokenExpiredException {
+        val resetToken = passwordService.findByToken(token)
+                .orElseThrow(() -> new NotFoundException("Provided token '" + token + "' not found"));
+
+        if (resetToken.getExpiryTimestamp().isBefore(LocalDateTime.now()) || resetToken.isUsed()) {
+            throw new PasswordResetTokenExpiredException("Provided token '" + token + "' expired");
+        }
+
+        return ResponseEntity.ok().build();
+    }
+
+    @ApiOperation("Reset password")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "Password reset was successful"),
+            @ApiResponse(code = 404, message = "Passwords were incorrect"),
+            @ApiResponse(code = 404, message = "The token was not found")
+    })
+    @PostMapping(value = "/password-reset")
+    @ResponseBody
+    public ResponseEntity<?> resetPassword(@RequestBody @Valid PasswordResetRequest passwordReset, @RequestParam("token") String token) throws NotFoundException, BadRequestException {
+        val optionalResetPasswordToken = passwordService.findByToken(token)
+                .orElseThrow(() -> new NotFoundException("Provided token '" + token + "' not found"));
+        if (optionalResetPasswordToken.getAppUser().getPassword().equals(passwordReset.getOldPassword())) {
+            AppUser user = optionalResetPasswordToken.getAppUser();
+            user.setPassword(passwordReset.getNewPassword());
+            appUserService.update(user);
+            optionalResetPasswordToken.setUsed(true);
```

Comment:
Do you check if it was used when actually resetting the password? I only see the check in `/token-validate`.
Also, shouldn't the token just be deleted after it is used? Currently, if I understand the code correctly, PasswordService won't be able to generate a second token for an AppUser. Maybe user could generate another token even if he has one already?